### DISCUSSION TITLE
Batch 处理时，只会处理识别出的编号为 0 的框吗？

### DISCUSSION THREAD

**Discussion by OedoSoldier** [continue-revolution/sd-webui-segment-anything#61] [Q&A]
RT

**Comment by continue-revolution** [comment]
会处理所有边框，因为没有点，没有点才只能处理一个边框。

如果有bug麻烦发发图片，方便我debug

**Comment by OedoSoldier** [comment]
我看另一个 GroundingDINO 插件（https://github.com/NeoGraph-K/sd-webui-ddsd ）支持 prompt 使用逻辑表达，似乎是直接把 prompt 分开处理了（https://github.com/NeoGraph-K/sd-webui-ddsd/blob/d5312d49bb5a5d86bd5aa925ca9a67a06491427f/scripts/ddsd_utils.py ），我觉得可以借鉴一下

**Comment by continue-revolution** [comment]
我看了一下，意思是不是比如说我girl AND tree就会把gril和tree同时框出来，然后比如说girl NOT eyes就会把girl除开眼睛以外的区域抠出来，如果是那我感觉应该很简单，可以在#57 之后支持一下

**Comment by continue-revolution** [comment]
之前一直没人提这个需求

**Comment by OedoSoldier** [comment]
😂我在抠初音的时候那条大马尾一动起来就容易脱锁，然后我用笨方法跑了两边，一遍跑girl一遍跑twintails，再合在一起

**Comment by continue-revolution** [comment]
可以，等我把这个大更新搞好了就来搞你这个，（也可能把你要的这个feature加入这次大更新中）

**Comment by continue-revolution** [comment]
你能不能发一下你那个miku的图，我用来做测试的

**Reply by OedoSoldier** [reply]
> 你能不能发一下你那个miku的图，我用来做测试的

https://i.imgur.com/6Iwktvq.png

**Reply by continue-revolution** [reply]
我感觉加一个类似于OR的运算符是不是就够了，就比如说你这种需求。感觉太复杂了好像也没什么应用场景

**Reply by continue-revolution** [reply]
https://github.com/IDEA-Research/GroundingDINO#star-explanationstips-for-grounding-dino-inputs-and-outputs
groundingdino似乎原生支持你的需求，那既然这样我就不做了

**Reply by continue-revolution** [reply]
![image](https://user-images.githubusercontent.com/63914308/233859131-d4dd5cfa-2b37-4745-b6a6-d7dd1f88d0ea.png)
![image](https://user-images.githubusercontent.com/63914308/233859145-1af57683-057a-44e3-999b-87719655d39f.png)
![image](https://user-images.githubusercontent.com/63914308/233859160-76dddb0e-efb4-44bc-82b7-df3f0bf7831c.png)

**Reply by OedoSoldier** [reply]
主要是批处理的时候，因为只处理第0个框，所以会有这种问题

**Reply by continue-revolution** [reply]
批处理我这儿没问题，你的text prompt可以输入girl . twintails试试

**Reply by OedoSoldier** [reply]
> 批处理我这儿没问题，你的text prompt可以输入girl . twintails试试

这张图没问题，但是本来正常的图可能会只把双马尾切出来……

**Reply by continue-revolution** [reply]
逆天，发发本来正常结果只切出双马尾的图吧，我看看框出啥了

**Reply by OedoSoldier** [reply]
https://i.imgur.com/1NHa6Wc.png

**Reply by continue-revolution** [reply]
![1NHa6Wc_2_output](https://user-images.githubusercontent.com/63914308/233879866-6769691a-b2de-4374-b560-a7188ddafc2c.png)

我这儿仍然是好的诶，你可能需要输出三张，然后自己筛一下。因为我的经验是排名第一的mask通常是performance比较差的...但是也不好说

**Reply by continue-revolution** [reply]
而且即便是第一张图也不是只有双马尾，好奇怪。或者说你是不是用了小模型，还是说你调了box threshold
![image](https://user-images.githubusercontent.com/63914308/233880129-8dd525d3-c11a-44e6-98f6-41118675de8d.png)

**Reply by continue-revolution** [reply]
你可以用我的参数，我这个虽然大更新过，但是没有更新批处理

**Reply by continue-revolution** [reply]
![Untitled](https://user-images.githubusercontent.com/63914308/233880254-482953f2-2a4c-4260-adf7-2ee20f55a485.png)
我的bounding box长这样，你可以看看你的bounding box长啥样

**Reply by continue-revolution** [reply]
threshold调成0.5之后就只剩下双马尾了

**Reply by OedoSoldier** [reply]
我仔细想了这个问题其实也没啥好的解决方法，就用不同的prompt跑两遍再合并是最佳的……

**Reply by continue-revolution** [reply]
先调低box threshold吧，至少我用0.3能work。
两个prompt求或应该很容易搞，应该很快就可以做出来，但是可能也正如你所说，仍然没那么稳定。别的什么与非异或之类的就需要重新用SAM出mask，有点烦人，不一定会做了。
新feature可能今晚/明天凌晨就要发布了，你可以试一下语义分割，我用你的图试过语义分割，我觉得还比较好用。
